docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,316 @@
+libraries so the issue of support should be solved with this contract.
+
+- Each provider may want to implement their own templating standard. This
+  could affect the UX of clusterctl when automatically pulling workload
+  cluster templates from various providers.
+
+
+### Constraints
+- Currently, clusterctl relies on the conformance of file name conventions
+  such as `infrastructure-components.yaml` and
+  `cluster-template-<flavor>.yaml`. Other templating tools might require other
+  conventions to be defined and followed to allow the same “day 1” experience.
+- Some templating tools will require multiple files to be defined rather than
+  a single yaml file. These artifacts will need to be “grouped” together to
+  support current retrieval mechanisms. Currently, `clusterctl config cluster`
+  retrieves templates from multiple sources such as ConfigMaps within a
+  cluster, URL, Github Repository, Local Repository and even the overrides
+  directory. To ensure compatibility, we’ll need to establish a compression
+  format like tar, gzip or zip.
+
+## Alternatives
+- An alternative to allowing clusterctl handle templates would be to just use
+  another tool upstream to generate the various templates with variables
+  defined. Then use clusterctl to consume that template and use the default
+  variable substitution templating solution.
+
+
+## Upgrade Strategy
+
+
+Any individual template engine libraries would be upgraded via go modules
+preferably using semver standards.
+
+
+## Additional Details
+
+
+Currently, cluster-api-provider-azure is using kustomize to author their
+cluster template flavors. See [issue 540] for more details. Maybe we can
+collect input from them to understand their strategy if they need to
+manage their template variations.
+
+
+## Implementation History
+
+
+- [wfernandes] 04/22/2020: Proposed idea in an issue or [community meeting]
+- [wfernandes/fabrizio.pandini] 04/27/2020: Compile a [CAEP Google Doc] following the CAEP template
+- [wfernandes] 05/06/2020: First round of feedback from community
+- [wfernandes] 05/11/2020: Open proposal PR
+- [wfernandes] 05/13/2020: Present proposal at a [community meeting]
+
+
+## Open Questions
+
+1. [wfernandes] We need to define a format or convention for retrieving multiple template files.
+    Currently, clusterctl supports template retrieval from a ConfigMap in a
+    cluster, URL, Github release artifacts and local file system. We default to
+    retrieving provider cluster templates from github release assets. In order to
+    avoid running into github quota limits on retrieving asset files we can pull a
+    compressed file with all the template contents.
+
+      As mentioned above in the *Constraints* and *Implementation* section a suggestion
+      for convention would be
+      `cluster-template-<template-engine-type>.<compression-format>`. Compression
+      format could be tar, gzip, zip, etc.
```

Comment:
This section should be removed before merging the PR, add it to risk/mitigations